src/geoserver-publish-imagemosaic/index.js

```diff
@@ -22,61 +22,91 @@ const grc = new GeoServerRestClient(url, user, pw);
+  const layerName = inputs[1];
   const coverageToAdd = inputs[2];
+  const replaceExistingGranule = inputs[3] ? inputs[3] : false;
+
   let newPath;
 
   const geoServerAvailable = await isGeoServerAvailable();
 
-  if (!geoServerAvailable ){
+  if (!geoServerAvailable) {
     console.log('Geoserver not available');
     console.log('Job should be requeued!');
     workerJob.missingPreconditions = true;
     return;
   }
 
   try {
-    // check if coverage store exists
-    const covStoreObject = await grc.datastores.getCoverageStore(workspace, covStore);
+    // Check if coverage store exists
+    const covStoreObject = await grc.datastores.getCoverageStore(ws, covStore);
 
     if (!covStoreObject) {
-      // TODO Add function to create new coverage store
-      // cf. grc.datastores.createImageMosaicStore
-      throw 'Datastore does not exist.'
+      throw `Coverage store ${covStore} does not exist.`;
     }
 
-    // move coverage to coverage store folder to keep file structure clean in data directory
+    // Move coverage to coverage store folder to keep file structure clean in data directory
     const fileName = path.basename(coverageToAdd);
-    const oldPath = coverageToAdd;
-    // the internal geoserver url always starts with 'file:', so we split it and use the second index
+    const dirName = path.dirname(coverageToAdd);
+    // The internal geoserver url always starts with 'file:', so we split it and use the second index
     const coverageStorePath = covStoreObject.coverageStore.url.split(":")[1];
-    // TODO read the geoserver data dir path from rest api
-    newPath = `/opt/geoserver_data/${coverageStorePath}/${fileName}`;
 
-    // Move geotiff
-    await fsPromises.rename(oldPath, newPath);
-    
+    newPath = `${dirName}/${coverageStorePath}/${fileName}`;
+    // Move GeoTiff
+    await fsPromises.rename(coverageToAdd, newPath);
+
+    // Check if granule already exists
+    const granules = await grc.imagemosaics.getGranules(ws, covStore, covStore);
+    const granuleAlreadyExists = granules.features.some(
+      feature => feature.properties.location === newPath
+    );
+
+    if (granuleAlreadyExists && !replaceExistingGranule) {
+      throw 'Granule with this timestamp already exists.';
+    }
+
     await grc.imagemosaics.addGranuleByServerFile(
-      workspace, covStore, newPath
+      ws, covStore, newPath
     );
     log('Successfully added new granule to coverage store.');
+
+    // check if layer has time dimension enabled
+    let hasTime = false;
+    const coverage = await grc.layers.getCoverage(ws, covStore, covStore);
+    if (coverage && coverage.coverage.metadata && coverage.coverage.metadata.entry &&
+      coverage.coverage.metadata.entry['@key'] === 'time' && (typeof coverage.coverage.metadata.entry.dimensionInfo === 'object')) {
+      const dimInfo = coverage.coverage.metadata.entry.dimensionInfo;
+      if (dimInfo.enabled === true && dimInfo.nearestMatchEnabled === true &&
```

Comment:
i think having nearest match enabled is optional and does not need to be checked